Case expression optimizer

cardillan opened the pull request:
Yet another optimizer for elimination of unnecessary assignments, this time for case expressions. The compiler stores the value of the case expression in an `__ast` variable, and this value is then compared to every where branch until a match is found. If the case expression itself is a variable, the `CaseExpressionOptimizer` will modify the where branches to use the original variable instead of the `__ast` variable in comparison instructions. The modification is quite safe, as it is checked that every instruction corresponding to the where branches matches the exact format generated by the compiler.

Note: the compiler can sometimes produce temporary variables with prefix `ast`. This could conflict with some user variable. As part of this change I've updated the prefix of these variables to `__ast`.

Commit:
Case expression optimizer

francois commented on compiler/src/main/java/info/teksol/mindcode/mindustry/optimisation/GlobalOptimizer.java:
```diff
@@ -61,7 +62,15 @@ protected int countInstructions(Predicate<LogicInstruction> matcher) {
+                program.set(index, replaced);
+                return;
+            }
+        }
```
I don't understand the difference between the two code blocks? I see that the original code was not checking the result of `indexOf`, which can return -1 (I think) for not found values.

At the same time, this is checking for object reference, not object equality. I think this is correct, but I wanted to flag this.

cardillan replied:
The reason is that LogicInstruction is a value object -- it has `equals` and `hashCode` methods based on its values. This means that two instructions can be identical even when they're two distinct instructions at different places in the program (eg. every `end` instruction is equal to every other). The previous version of the code therefore could end up replacing wrong instance of the instruction in the program, which resulted in one pretty mind-boggling bug :)

I'll add a comment to the method to explain the issue.

Btw, the DeadCodeEliminator also relies on selecting instructions for removal by their equality. It's not a problem there, because if an instruction is eligible for removal, so are all other copies of it. The subtlety was lost on me at first, though.

cardillan commented:
I've forgotten to move unit tests to different class when I moved the functionality for removing always-false jumps from the SingleStepJumpEliminator to the ConditionalJumpsNormalizer. I did it now and updated/force-pushed all the branches. Perhaps I shouldn't have - sorry if it is an inconvenience.
